parser.cc

```diff
@@ -1453,6 +1454,52 @@ Input::Index AsciiFileInput::size() const
 	return file_size;
 }
 
+StreamInput StreamInput::Create(std::istream& s)
+{
+	const size_t start = s.tellg();
+	s.seekg(0, std::ios::end);
+
+	const size_t len = static_cast<size_t>(s.tellg()) - start;
+	s.seekg(start, std::ios::beg);
+
+	return StreamInput(s, len);
+}
+
+StreamInput::StreamInput(std::istream& s, size_t len)
+	: length(len), stream(s)
+{
+}
+
+bool StreamInput::fillBuffer(Index start, Index &len, char32_t *&b)
+{
+	if (start > length)
+	{
+		return false;
+	}
+
+	char buffer[static_buffer_size];
+	len = std::min(this->length, static_buffer_size);
+	len = std::min(len, this->length - start);
+
+	stream.read(buffer, len);
```

Comment:
Unlike the file descriptor version, this will move the stream pointer around and so should probably take exclusive ownership of the stream (i.e be constructed by an r-value reference) and provide an accessor for getting the stream back after parsing for things that want to parse a bit with Pegmatite and then read the rest of the stream.